### PR TITLE
Generic Calendar Objects introduced (task #5394)

### PR DESCRIPTION
**Pull request by anvyst**
As a first major step for refactoring existing code, generic pool of objects is introduced to facilitate instance conversion between external libraries (aka Google Calendar, Office 365) to native Cake MVC models/controllers.

### CODE REVIEW

File: .editorconfig [hunk 1]
```diff
@@ -10,6 +10,14 @@ end_of_line = lf
 insert_final_newline = true
 trim_trailing_whitespace = true
 
+[*.{js,vue,json}]
```
Comment (mamchenkov) [inline .editorconfig]: @anvyst for the JSON we need to use an uglier style.  We have a number of configs throughout the system that uses JSON as a format, and we do automatic processing of those (saving the results of the `json_encode()` with PRETTY).  I think for that we need like 4 spaces and no end of line or something.

File: config/Migrations/20180308091250_AddEventTypesToCalendars.php [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+     * Change Method.
+     *
+     * More information on this method is available here:
+     * http://docs.phinx.org/en/latest/migrations.html#the-change-method
+     * @return void
+     */
+    public function change()
+    {
+        $table = $this->table('calendars');
+        $table->addColumn('event_types', 'text', [
```
Comment (mamchenkov) [inline config/Migrations/20180308091250_AddEventTypesToCalendars.php]: @anvyst I'm not convinced yet about storing JSON in the database long text field.  We need to discuss this.

File: config/Schema/calendar.json [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+      "$ref": "definitions.json#/field"
+    },
+    "active": {
+      "$ref": "definitions.json#/field"
+    },
+    "is_public": {
+      "$ref": "definitions.json#/field"
+    },
+    "editable": {
+      "$ref": "definitions.json#/field"
```
Comment (mamchenkov) [inline config/Schema/calendar.json]: @anvyst this is sort of defeats the purpose of having JSON Schema ... if all fields are defined as a very generic type that applies to everything.

File: src/Event/Plugin/GetCalendarsListener.php [hunk 4]
```diff
@@ -33,9 +37,80 @@ public function implementedEvents()
...
+        $entities = $result = [];
+
+        $table = $event->subject();
+
+        $map = ObjectFactory::getParserConfig($table->alias(), 'Event', $options->getArrayCopy());
+        $calendarsTable = TableRegistry::get('Qobo/Calendar.Calendars');
+
+        $calendars = $calendarsTable->getByAllowedEventTypes($table->alias());
+
+        if (!empty($calendars)) {
```
Comment (mamchenkov) [inline src/Event/Plugin/GetCalendarsListener.php]: @anvyst refactor into a separate method

File: src/Event/Plugin/GetCalendarsListener.php [hunk 5]
```diff
@@ -33,9 +37,80 @@ public function implementedEvents()
...
+
+                if (!$calendarEntity) {
+                    continue;
+                }
+
+                $entities[] = $calendarEntity;
+            }
+        }
+
+        if (!empty($entities)) {
```
Comment (mamchenkov) [inline src/Event/Plugin/GetCalendarsListener.php]: @anvyst refactor into a separate method

File: config/Schema/calendar.json (hunk 3 above)
Comment (anvyst) [reply config/Schema/calendar.json]: @mamchenkov the definition is quiet strict. It allows only a certain type of `field` value. 
It's either field (for direct mapping), value - for default values, and callback - for Model defined function call that would return customised result for the field.

It's flexible enough to meet current requirements

File: config/Migrations/20180308091250_AddEventTypesToCalendars.php (hunk 2 above)
Comment (anvyst) [reply config/Migrations/20180308091250_AddEventTypesToCalendars.php]: @mamchenkov `event_types` used just to store the aggregated list of event types used can create. One way - you can define calendar with its own event types. Another one - generic object whenever is stored as an event. For example: User upon being saved in the system can be transformed into `calendar_events/birthday.json` event.

Comment (mamchenkov) [reply config/Migrations/20180308091250_AddEventTypesToCalendars.php]: @anvyst we need to whiteboard this ... there are a few alarm bells ringing in my head. :)  One particular area of issues is around de-synchronization of the configuration files with the values stored in the database.

File: config/Schema/calendar.json (hunk 3 above)
Comment (mamchenkov) [reply config/Schema/calendar.json]: @anvyst the JSON schema definition is not strict here at all.  As it allows all combinations between the type and value.  For example, setting type to callback and value to number will be valid.  If you want to allow only particular combinations, then different definitions should be used and/or the anyOf, allOff, oneOf rules.

**Pull request merged**